src/Lucene.Net/Util/LightThreadLocal.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Lucene.Net.Store;
+
+namespace Lucene.Net.Util
+{
+    internal class LightThreadLocal<T> : IDisposable
+    {
+        [ThreadStatic] private static CurrentThreadState _state;
+        private ConcurrentDictionary<CurrentThreadState, T> _values = new ConcurrentDictionary<CurrentThreadState, T>(ReferenceComparer<CurrentThreadState>.Instance);
+
+        private readonly Func<IState, T> _generator;
+
+        public LightThreadLocal(Func<IState, T> generator = null)
+        {
+            _generator = generator;
+        }
+
+        public void Set(T value)
+        {
+            (_state ??= new CurrentThreadState()).Register(this);
+            _values[_state] = value;
+        }
+
+        public T Get(IState state = null)
+        {
+            (_state ??= new CurrentThreadState()).Register(this);
+            if (_values.TryGetValue(_state, out var v) == false
+                && _generator != null)
+            {
+                v = _generator(state);
+                _values[_state] = v;
+            }
+
+            return v;
+        }
+
+        private sealed class ReferenceComparer<T> : IEqualityComparer<T>
+        {
+            public static readonly ReferenceComparer<T> Instance = new ReferenceComparer<T>();
+            public bool Equals(T x, T y)
+            {
+                return ReferenceEquals(x, y);
+            }
+
+            public int GetHashCode(T obj)
+            {
+                return RuntimeHelpers.GetHashCode(obj);
+            }
+        }
+
+        public class CurrentThreadState
+        {
+            private readonly HashSet<LightThreadLocal<T>> _parents = new HashSet<LightThreadLocal<T>>(ReferenceComparer<LightThreadLocal<T>>.Instance);
+            public void Register(LightThreadLocal<T> parent)
+            {
+                _parents.Add(parent);

```

Comment:
This is allowed to be run concurrently, isn't it? Then shouldn't we use `ConcurrentBag` instead of `HashSet`?

Reply:
Hm.. but now I see it needs to ensure uniqueness. Sill - is it safe to use regular `HashSet` here?

Reply:
This is thread static value, no concurrency possible